core/src/main/java/org/infinispan/statetransfer/CommitManager.java

```diff
@@ -68,15 +68,30 @@ public final void stopTrack(Flag track) {
          }
          tracker.clear();
       } else {
-         for (Iterator<Map.Entry<Object, DiscardPolicy>> iterator = tracker.entrySet().iterator();
-              iterator.hasNext(); ) {
-            if (iterator.next().getValue().update(trackStateTransfer, trackXSiteStateTransfer)) {
-               iterator.remove();
+         for (Map.Entry<Integer, Map<Object, DiscardPolicy>> segment : tracker.entrySet()) {
+            Map<Object, DiscardPolicy> entries = segment.getValue();
+            entries.entrySet()
+                  .removeIf(policy -> policy.getValue().update(trackStateTransfer, trackXSiteStateTransfer));
+
+            if (entries.isEmpty()) {
+               tracker.remove(segment.getKey());
             }
          }
       }
    }
 
+   /**
+    * Stop tracking the given entries if the tracking is enabled.
+    *
+    * @param flag: flag to verify if tracking is enabled.
+    * @param segmentId: segment to stop tracking.
+    */
+   public final void stopTrackFor(Flag flag, int segmentId) {
```

Comment:
@pruivo You may have to comment here about xsite state transfer if this is okay or not.

Reply:
We cannot optimize for cross-site state transfer because we don't enforce the same hashing function / number of segments / number of owners. 
A key may belong to segment 1 in "site_a" but it may belong to segment 122 in "site_b".

I don't think this method implementation is correct. If you have a cross-site state transfer and a "local cluster" state transfer, you cannot remove the segment from the `tracker` map. You need to stop tracking for the "local cluter" state transfer only.

Reply:
Added a verification for the xsite flag, we only remove entries for state transfer and keep the xsite ones, working in a similar way as the `stopTrack(Flag track)`. Added a test for this case as well, handling state transfer with the xsite transfers simultaneously.